Fails to build on Linux 2.6.31

```
[root@segv-fc12 src]# make
if [ ! -f kernel-config.h ] || tail -1 kernel-config.h | grep -qv '#endif'; then ./genconfig.sh 2.6.31.5-127.fc12.x86_64; fi;
make -C /lib/modules/2.6.31.5-127.fc12.x86_64/build M=/root/dattobd/src modules
make[1]: Entering directory `/usr/src/kernels/2.6.31.5-127.fc12.x86_64'
  CC [M]  /root/dattobd/src/dattobd.o
/root/dattobd/src/dattobd.c:354: error: conflicting types for ‘blk_stack_limits’
include/linux/blkdev.h:920: note: previous declaration of ‘blk_stack_limits’ was here
/root/dattobd/src/dattobd.c: In function ‘blk_stack_limits’:
/root/dattobd/src/dattobd.c:357: error: ‘struct request_queue’ has no member named ‘max_sectors’
/root/dattobd/src/dattobd.c:357: error: ‘struct request_queue’ has no member named ‘max_sectors’
/root/dattobd/src/dattobd.c:357: error: ‘struct request_queue’ has no member named ‘max_sectors’
/root/dattobd/src/dattobd.c:357: error: ‘struct request_queue’ has no member named ‘max_sectors’
/root/dattobd/src/dattobd.c:357: error: ‘struct request_queue’ has no member named ‘max_sectors’
/root/dattobd/src/dattobd.c:357: error: ‘struct request_queue’ has no member named ‘max_sectors’
/root/dattobd/src/dattobd.c:357: warning: type defaults to ‘int’ in declaration of ‘_min1’
/root/dattobd/src/dattobd.c:357: error: ‘struct request_queue’ has no member named ‘max_sectors’
/root/dattobd/src/dattobd.c:357: error: ‘struct request_queue’ has no member named ‘max_sectors’
/root/dattobd/src/dattobd.c:357: warning: type defaults to ‘int’ in declaration of ‘_min2’
/root/dattobd/src/dattobd.c:357: error: ‘struct request_queue’ has no member named ‘max_sectors’
/root/dattobd/src/dattobd.c:358: error: ‘struct request_queue’ has no member named ‘max_hw_sectors’
/root/dattobd/src/dattobd.c:358: error: ‘struct request_queue’ has no member named ‘max_hw_sectors’
/root/dattobd/src/dattobd.c:358: error: ‘struct request_queue’ has no member named ‘max_hw_sectors’
/root/dattobd/src/dattobd.c:358: error: ‘struct request_queue’ has no member named ‘max_hw_sectors’
/root/dattobd/src/dattobd.c:358: error: ‘struct request_queue’ has no member named ‘max_hw_sectors’
/root/dattobd/src/dattobd.c:358: error: ‘struct request_queue’ has no member named ‘max_hw_sectors’
/root/dattobd/src/dattobd.c:358: warning: type defaults to ‘int’ in declaration of ‘_min1’
/root/dattobd/src/dattobd.c:358: error: ‘struct request_queue’ has no member named ‘max_hw_sectors’
/root/dattobd/src/dattobd.c:358: error: ‘struct request_queue’ has no member named ‘max_hw_sectors’
/root/dattobd/src/dattobd.c:358: warning: type defaults to ‘int’ in declaration of ‘_min2’
/root/dattobd/src/dattobd.c:358: error: ‘struct request_queue’ has no member named ‘max_hw_sectors’
/root/dattobd/src/dattobd.c:359: error: ‘struct request_queue’ has no member named ‘bounce_pfn’
/root/dattobd/src/dattobd.c:359: error: ‘struct request_queue’ has no member named ‘bounce_pfn’
/root/dattobd/src/dattobd.c:359: error: ‘struct request_queue’ has no member named ‘bounce_pfn’
/root/dattobd/src/dattobd.c:359: error: ‘struct request_queue’ has no member named ‘bounce_pfn’
/root/dattobd/src/dattobd.c:359: error: ‘struct request_queue’ has no member named ‘bounce_pfn’
/root/dattobd/src/dattobd.c:359: error: ‘struct request_queue’ has no member named ‘bounce_pfn’
/root/dattobd/src/dattobd.c:359: warning: type defaults to ‘int’ in declaration of ‘_min1’
/root/dattobd/src/dattobd.c:359: error: ‘struct request_queue’ has no member named ‘bounce_pfn’
/root/dattobd/src/dattobd.c:359: error: ‘struct request_queue’ has no member named ‘bounce_pfn’
/root/dattobd/src/dattobd.c:359: warning: type defaults to ‘int’ in declaration of ‘_min2’
/root/dattobd/src/dattobd.c:359: error: ‘struct request_queue’ has no member named ‘bounce_pfn’
/root/dattobd/src/dattobd.c:360: error: ‘struct request_queue’ has no member named ‘seg_boundary_mask’
/root/dattobd/src/dattobd.c:360: error: ‘struct request_queue’ has no member named ‘seg_boundary_mask’
/root/dattobd/src/dattobd.c:360: error: ‘struct request_queue’ has no member named ‘seg_boundary_mask’
/root/dattobd/src/dattobd.c:360: error: ‘struct request_queue’ has no member named ‘seg_boundary_mask’
/root/dattobd/src/dattobd.c:360: error: ‘struct request_queue’ has no member named ‘seg_boundary_mask’
/root/dattobd/src/dattobd.c:360: error: ‘struct request_queue’ has no member named ‘seg_boundary_mask’
/root/dattobd/src/dattobd.c:360: warning: type defaults to ‘int’ in declaration of ‘_min1’
/root/dattobd/src/dattobd.c:360: error: ‘struct request_queue’ has no member named ‘seg_boundary_mask’
/root/dattobd/src/dattobd.c:360: error: ‘struct request_queue’ has no member named ‘seg_boundary_mask’
/root/dattobd/src/dattobd.c:360: warning: type defaults to ‘int’ in declaration of ‘_min2’
/root/dattobd/src/dattobd.c:360: error: ‘struct request_queue’ has no member named ‘seg_boundary_mask’
/root/dattobd/src/dattobd.c:362: error: ‘struct request_queue’ has no member named ‘max_phys_segments’
/root/dattobd/src/dattobd.c:362: error: ‘struct request_queue’ has no member named ‘max_phys_segments’
/root/dattobd/src/dattobd.c:362: error: ‘struct request_queue’ has no member named ‘max_phys_segments’
/root/dattobd/src/dattobd.c:362: error: ‘struct request_queue’ has no member named ‘max_phys_segments’
/root/dattobd/src/dattobd.c:362: error: ‘struct request_queue’ has no member named ‘max_phys_segments’
/root/dattobd/src/dattobd.c:362: error: ‘struct request_queue’ has no member named ‘max_phys_segments’
/root/dattobd/src/dattobd.c:362: warning: type defaults to ‘int’ in declaration of ‘_min1’
/root/dattobd/src/dattobd.c:362: error: ‘struct request_queue’ has no member named ‘max_phys_segments’
/root/dattobd/src/dattobd.c:362: error: ‘struct request_queue’ has no member named ‘max_phys_segments’
/root/dattobd/src/dattobd.c:362: warning: type defaults to ‘int’ in declaration of ‘_min2’
/root/dattobd/src/dattobd.c:362: error: ‘struct request_queue’ has no member named ‘max_phys_segments’
/root/dattobd/src/dattobd.c:364: error: ‘struct request_queue’ has no member named ‘max_hw_segments’
/root/dattobd/src/dattobd.c:364: error: ‘struct request_queue’ has no member named ‘max_hw_segments’
/root/dattobd/src/dattobd.c:364: error: ‘struct request_queue’ has no member named ‘max_hw_segments’
/root/dattobd/src/dattobd.c:364: error: ‘struct request_queue’ has no member named ‘max_hw_segments’
/root/dattobd/src/dattobd.c:364: error: ‘struct request_queue’ has no member named ‘max_hw_segments’
/root/dattobd/src/dattobd.c:364: error: ‘struct request_queue’ has no member named ‘max_hw_segments’
/root/dattobd/src/dattobd.c:364: warning: type defaults to ‘int’ in declaration of ‘_min1’
/root/dattobd/src/dattobd.c:364: error: ‘struct request_queue’ has no member named ‘max_hw_segments’
/root/dattobd/src/dattobd.c:364: error: ‘struct request_queue’ has no member named ‘max_hw_segments’
/root/dattobd/src/dattobd.c:364: warning: type defaults to ‘int’ in declaration of ‘_min2’
/root/dattobd/src/dattobd.c:364: error: ‘struct request_queue’ has no member named ‘max_hw_segments’
/root/dattobd/src/dattobd.c:366: error: ‘struct request_queue’ has no member named ‘max_segment_size’
/root/dattobd/src/dattobd.c:366: error: ‘struct request_queue’ has no member named ‘max_segment_size’
/root/dattobd/src/dattobd.c:366: error: ‘struct request_queue’ has no member named ‘max_segment_size’
/root/dattobd/src/dattobd.c:366: error: ‘struct request_queue’ has no member named ‘max_segment_size’
/root/dattobd/src/dattobd.c:366: error: ‘struct request_queue’ has no member named ‘max_segment_size’
/root/dattobd/src/dattobd.c:366: error: ‘struct request_queue’ has no member named ‘max_segment_size’
/root/dattobd/src/dattobd.c:366: warning: type defaults to ‘int’ in declaration of ‘_min1’
/root/dattobd/src/dattobd.c:366: error: ‘struct request_queue’ has no member named ‘max_segment_size’
/root/dattobd/src/dattobd.c:366: error: ‘struct request_queue’ has no member named ‘max_segment_size’
/root/dattobd/src/dattobd.c:366: warning: type defaults to ‘int’ in declaration of ‘_min2’
/root/dattobd/src/dattobd.c:366: error: ‘struct request_queue’ has no member named ‘max_segment_size’
make[2]: *** [/root/dattobd/src/dattobd.o] Error 1
make[1]: *** [_module_/root/dattobd/src] Error 2
make[1]: Leaving directory `/usr/src/kernels/2.6.31.5-127.fc12.x86_64'
make: *** [default] Error 2
```

--- a/src/dattobd.c
+++ b/src/dattobd.c
@@ -347,36 +347,42 @@ void path_put(const struct path *path) {
 #define UMOUNT_NOFOLLOW 0
 #endif
 
-#ifndef HAVE_BDEV_STACK_LIMITS
+#if !defined(HAVE_BDEV_STACK_LIMITS) && !defined(HAVE_BLK_SET_DEFAULT_LIMITS)
+//#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,31)
+
 #ifndef min_not_zero
 #define min_not_zero(l, r) (l == 0) ? r : ((r == 0) ? l : min(l, r))
 #endif
-int blk_stack_limits(struct request_queue *t, struct request_queue *b,
-		     sector_t offset)
-{
+
+int blk_stack_limits(struct request_queue *t, struct request_queue *b, sector_t offset){
 	t->max_sectors = min_not_zero(t->max_sectors, b->max_sectors);
 	t->max_hw_sectors = min_not_zero(t->max_hw_sectors, b->max_hw_sectors);
 	t->bounce_pfn = min_not_zero(t->bounce_pfn, b->bounce_pfn);
-	t->seg_boundary_mask = min_not_zero(t->seg_boundary_mask,
-					    b->seg_boundary_mask);
-	t->max_phys_segments = min_not_zero(t->max_phys_segments,
-					    b->max_phys_segments);
-	t->max_hw_segments = min_not_zero(t->max_hw_segments,
-					  b->max_hw_segments);
-	t->max_segment_size = min_not_zero(t->max_segment_size,
-					   b->max_segment_size);
+	t->seg_boundary_mask = min_not_zero(t->seg_boundary_mask, b->seg_boundary_mask);
+	t->max_phys_segments = min_not_zero(t->max_phys_segments, b->max_phys_segments);
+	t->max_hw_segments = min_not_zero(t->max_hw_segments, b->max_hw_segments);
+	t->max_segment_size = min_not_zero(t->max_segment_size, b->max_segment_size);
 	return 0;
 }
-int bdev_stack_limits(struct request_queue *t, struct block_device *bdev, sector_t start)
-{
+
+int dattobd_bdev_stack_limits(struct request_queue *t, struct block_device *bdev, sector_t start){
 	struct request_queue *bq = bdev_get_queue(bdev);
 	start += get_start_sect(bdev);
 	return blk_stack_limits(t, bq, start << 9);
 }
 
-#define dattobd_bdev_stack_limits(queue, bdev, start)
+#elif !defined(HAVE_BDEV_STACK_LIMITS)
+//#elif LINUX_VERSION_CODE < KERNEL_VERSION(2,6,32)
+
+int bdev_stack_limits(struct queue_limits *t, struct block_device *bdev, sector_t start){
+	struct request_queue *bq = bdev_get_queue(bdev);
+	start += get_start_sect(bdev);
+	return blk_stack_limits(t, bq, start << 9);
+}
+#define dattobd_bdev_stack_limits(queue, bdev, start) bdev_stack_limits(&(queue)->limits, bdev, start)
+
 #else
-#define dattobd_bdev_stack_limits(queue, bdev, start) bdev_stack_limits(queue->limits, bdev, start)
+#define dattobd_bdev_stack_limits(queue, bdev, start) bdev_stack_limits(&(queue)->limits, bdev, start)
 #endif
 
 #ifndef HAVE_KERN_PATH
@@ -3355,7 +3361,7 @@ static int __tracer_setup_snap(struct snap_device *dev, unsigned int minor, stru
 	//give our request queue the same properties as the base device's
 	LOG_DEBUG("setting queue limits");
 	blk_set_stacking_limits(&dev->sd_queue->limits);
-	dattobd_bdev_stack_limits(&dev->sd_queue, bdev, 0);
+	dattobd_bdev_stack_limits(dev->sd_queue, bdev, 0);
 
 #ifdef HAVE_MERGE_BVEC_FN
 	//use a thin wrapper around the base device's merge_bvec_fn

